File: examples/nodejs/with-events.js
/**
 * Conversion with Event Listeners
 *
 * This example shows how to use event listeners to track conversion progress
 */

const ConvertorioClient = require('../../libs/nodejs');

const client = new ConvertorioClient({
    apiKey: 'your_api_key_here'
});

// Listen to conversion events
client.on('start', (data) => {
    console.log(`🚀 Starting conversion: ${data.fileName}`);
    console.log(`   ${data.sourceFormat.toUpperCase()} → ${data.targetFormat.toUpperCase()}\n`);
});

client.on('progress', (data) => {
    console.log(`⏳ ${data.message}`);
    if (data.jobId) {
        console.log(`   Job ID: ${data.jobId}`);
    }
});

client.on('status', (data) => {
    console.log(`📊 Status: ${data.status} (attempt ${data.attempt}/${data.maxAttempts})`);
});

client.on('complete', (data) => {
    console.log('\n✅ Conversion completed!');
    console.log('📁 Output file:', data.outputPath);
    console.log('⚡ Processing time:', `${data.processingTime} ms`);
    console.log('💾 File size:', `${(data.fileSize / 1024).toFixed(2)} KB`);
});

client.on('error', (data) => {
    console.error('\n❌ Error:', data.error);
});

async function main() {
    try {
        await client.convertFile({
            inputPath: './input.png',
            targetFormat: 'webp',
            outputPath: './output.webp'
        });
    } catch (error) {
        console.error('Conversion failed:', error.message);
        process.exit(1);
    }
}

main();

File: examples/ruby/with_events.rb
#!/usr/bin/env ruby

# Conversion with Event Listeners
#
# This example shows how to use event listeners to track conversion progress

require_relative '../../libs/ruby/lib/convertorio'

client = Convertorio::Client.new(
  api_key: 'your_api_key_here'
)

# Listen to conversion events
client.on(:start) do |data|
  puts "🚀 Starting conversion: #{data[:file_name]}"
  puts "   #{data[:source_format].upcase} → #{data[:target_format].upcase}\n\n"
end

client.on(:progress) do |data|
  puts "⏳ #{data[:message]}"
  puts "   Job ID: #{data[:job_id]}" if data[:job_id]
end

client.on(:status) do |data|
  puts "📊 Status: #{data[:status]} (attempt #{data[:attempt]}/#{data[:max_attempts]})"
end

client.on(:complete) do |data|
  puts "\n✅ Conversion completed!"
  puts "📁 Output file: #{data[:output_path]}"
  puts "⚡ Processing time: #{data[:processing_time]} ms"
  puts "💾 File size: #{'%.2f' % (data[:file_size] / 1024.0)} KB"
end

client.on(:error) do |data|
  puts "\n❌ Error: #{data[:error]}"
end

begin
  client.convert_file(
    input_path: './input.png',
    target_format: 'webp',
    output_path: './output.webp'
  )
rescue Convertorio::Error => e
  puts "Conversion failed: #{e.message}"
  exit(1)
end
